gstcefsrc.cc

```diff
@@ -525,66 +654,111 @@ run_cef (GstCefSrc *src)
     /* unblock start () */
     g_mutex_lock (&init_lock);
     cef_inited = TRUE;
-    g_cond_signal(&init_cond);
+    g_cond_broadcast (&init_cond);
     g_mutex_unlock (&init_lock);
 
-    goto done;
+    return NULL;
   }
 
   g_mutex_lock (&init_lock);
   cef_inited = TRUE;
   init_result = TRUE;
-  g_cond_signal(&init_cond);
+  cef_shutdown = FALSE;
+  g_cond_broadcast (&init_cond);
   g_mutex_unlock (&init_lock);
 
+#ifndef __APPLE__
   CefRunMessageLoop();
-
-  CefShutdown();
-
-  g_mutex_lock (&init_lock);
-  cef_inited = FALSE;
-  g_cond_signal(&init_cond);
-  g_mutex_unlock (&init_lock);
-
-done:
+  gst_cef_shutdown(nullptr);
+#endif
   return NULL;
 }
 
-void quit_message_loop (int arg)
+static GstStateChangeReturn
+gst_cef_handle_lifetime(GstElement *element, GstStateChange transition)
 {
-  CefQuitMessageLoop();
-}
-
-class ShutdownEnforcer {
- public:
-  ~ShutdownEnforcer() {
-    if (!cef_inited)
-      return;
+  auto result = GST_CALL_PARENT_WITH_DEFAULT (GST_ELEMENT_CLASS , change_state, (element, transition), GST_STATE_CHANGE_FAILURE);
 
-    CefPostTask(TID_UI, base::BindOnce(&quit_message_loop, 0));
-
-    g_mutex_lock(&init_lock);
-    while (cef_inited)
+  switch(transition)
+  {
+  case GST_STATE_CHANGE_NULL_TO_READY:
+  {
+    g_mutex_lock (&init_lock);
+    // Wait till a previous CEF is dismantled or completes initialization
+    while (cef_shutdown || cef_initializing)
       g_cond_wait (&init_cond, &init_lock);
+    if (cef_inited && !init_result) {
+      // BAIL OUT, CEF is not loaded.
+      result = GST_STATE_CHANGE_FAILURE;
+    } else if (!cef_inited) {
+      cef_initializing = TRUE;
+      g_cond_broadcast (&init_cond);
```

Comment:
That's not wrong, but I don't think anything is waiting for `cef_initializing = TRUE` is there?

Reply:
If multiple pipelines are created simultaneously with CEF down, only one must run the initializer. `!cef_inited` is enough to check for initialization completion, but in the case of thread-based initialization, I think a second check is needed to prevent multiple threads.

Reply:
hm, this isn't a check though, this is simply notifying other threads that the status has moved to "Initializing" :)

In practice, the broadcast that will matter here is the one done after the status has been set to "initialized", either from this thread or from another one.

Reply:
I've removed this broadcast in favour of the one produced by the initialization routine.